### PR TITLE
feat: DPS310 driver

### PR DESCRIPTION
**Pull request by trombik**
Requested at https://github.com/UncleRus/esp-idf-lib/issues/456

Sponserd by beriberikix.

Implemented I2C digital interface.

### CODE REVIEW

File: components/dps310/src/dps310.c [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+#include "helper_i2c.h"
+
+/* public headers */
+#include "dps310.h"
+
+static const char *TAG = "dps310";
+
+/* see 4.9.3 Compensation Scale Factors */
+#define N_SCALE_FACTORS (8)
+const int32_t scale_factors[N_SCALE_FACTORS] = {
```
Comment (UncleRus) [inline components/dps310/src/dps310.c:46]: static

File: components/dps310/src/dps310.c [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+}
+
+esp_err_t dps310_get_pm_rate(dps310_t *dev, uint8_t *value)
+{
+    uint8_t reg_value = 0;
+    esp_err_t err = ESP_FAIL;
+
+    CHECK_ARG(dev && value);
+
+    I2C_DEV_TAKE_MUTEX(&dev->i2c_dev);
```
Comment (UncleRus) [inline components/dps310/src/dps310.c:214]: maybe `_read_reg_mask()`

File: components/dps310/src/helper_i2c.c [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        }
+    }
+    else
+    {
+        count = 8;
+    }
+    return count;
+}
+
+esp_err_t _read_reg_nolock(i2c_dev_t *dev, uint8_t reg, uint8_t *val)
```
Comment (UncleRus) [inline components/dps310/src/helper_i2c.c:59]: inline

File: components/dps310/include/dps310.h [hunk 4]
```diff
@@ -0,0 +1,789 @@
...
+#define DPS310_REG_SENSOR_VALUE_LEN (3)
+
+/* See 8.9 Soft Reset and FIFO flush (RESET) */
+#define DPS310_FIFO_FLUSH_VALUE (1 << 7)
+#define DPS310_SOFT_RST_VALUE   (0b1001)
+
+/**
+ * @brief Initialize device descriptor
+ *
+ * The device descriptor is dynamically allocated by the driver, and keeps its
```
Comment (trombik) [inline components/dps310/include/dps310.h:355]: this is a lie.

File: components/dps310/include/dps310.h [hunk 5]
```diff
@@ -0,0 +1,789 @@
...
+ *
+ * @param[in] dev Device descriptor
+ * @return `ESP_OK` on success, `ESP_ERR_INVALID_ARG` when `dev` and/or `config` is NULL, or other errors when I2C communication fails.
+ */
+esp_err_t dps310_reset(dps310_t *dev);
+
+/**
+ * @brief Get pressure measurement rate.
+ *
+ * @param[in] dev pointer to the device descriptor
```
Comment (trombik) [inline components/dps310/include/dps310.h:412]: `*dev` is a pointer, but `dev` itself is not.

File: components/dps310/include/dps310.h [hunk 6]
```diff
@@ -0,0 +1,789 @@
...
+
+/**
+ * @brief Get pressure oversampling rate.
+ *
+ * @param[in] dev pointer to the device descriptor
+ * @param[out] value the value in the resister
+ * @return `ESP_OK` on success. `ESP_ERR_INVALID_ARG` when `dev` and/or
+ * `value` is NULL, or other errors when I2C communication fails.
+ */
+esp_err_t dps310_get_pm_prc(dps310_t *dev, uint8_t *value);
```
Comment (trombik) [inline components/dps310/include/dps310.h:455]: i really don't like `pm_prc`, which is what the datasheet called. `p_oversampling` might be better.

File: components/dps310/priv_include/helper_i2c.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+#include <i2cdev.h>
+
+/* private headers */
+#include "helper_macro.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+esp_err_t _read_reg_nolock(i2c_dev_t *dev, uint8_t reg, uint8_t *val);
```
Comment (trombik) [inline components/dps310/priv_include/helper_i2c.h:34]: document what it does briefly (at the least).

File: components/dps310/src/dps310.c (hunk 2 above)
Comment (trombik) [reply components/dps310/src/dps310.c:318]: should be. this one is the first function  i wrote for the driver and later i wrote `_read_reg_mask`.

File: components/dps310/src/helper_i2c.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+#include "helper_i2c.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static const char *TAG = "dps310/helper_i2c";
+
+/* count the number of trailing zero in a value, usually bitmasks. */
+static uint8_t count_trailing_zero_bits(uint8_t v)
```
Comment (trombik) [inline components/dps310/src/helper_i2c.c:41]: prefix it with `_` (and document what it means).

File: examples/dps310/default/README.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# Example for `dps310` driver
+
+## What it does
+
+The example application initializes `DPS310` device. In a loop, it reads
+the measured pressure and temperature values, and print them.
```
Comment (trombik) [inline examples/dps310/default/README.md:6]: this is a lie.

**Comment by beriberikix** [comment]
I was able to test the sample on a ESP32-S3 & Adafruit sensor board and had no issues!

**Comment by trombik** [comment]
@beriberikix the public APIs are considered as stable now. please have a look at the public header, and the example. I would like to request for your comment, specifically; do the naming conventions in functions, enum, and typedef look clear? I kept original words where appropriate, but modified some (such as `dps310_tmp_oversampling_t`)

I will continue to work on background mode and interrupt, but the driver is now usable for simple use cases. SPI mode is a bit complicated because incompatible SPI APIs in ESP8266 RTOS SDK and esp-idf. the SPI interface is probably designed for use cases in which performance is critical (such as drones). but command mode is enough for your use case, i guess (and if you really need it, it is possible to implement your own by using public APIs).

**Comment by trombik** [comment]
@UncleRus I believe that everything you mentioned is fixed. if you are okay with the PR. please merge.

I have already noticed that some functions do not have necessary `CHECK_ARG()`, but they have been fixed in another branch, which I will submit as PR later.

**Comment by UncleRus** [comment]
Thank you!

**Pull request merged**